ras_party/controllers/share_survey_controller.py

```diff
@@ -77,7 +85,95 @@ def get_unique_pending_shares(session):
+        raise Conflict("Expired share survey token")
+    except (BadSignature, BadData):
+        logger.exception("Bad token in validate_share_survey_token")
+        raise NotFound("Unknown batch number in token")
+    share_surveys = query_share_survey_by_batch_no(batch_no, session)
+    if len(share_surveys) == 0:
+        raise NotFound('Batch number does not exist')
+    return [share_survey.to_share_dict() for share_survey in share_surveys]
+
+
+@with_db_session
+def accept_share_survey(batch_no, session):
+    """
+    Confirms share surveys
+    Creates Enrolment records
+    Business Respondent records
+    Removes pending shares
+    :param: batch_no
+    :param: session
+    """
+    logger.info('Attempting to confirm pending share survey', batch_no=batch_no)
+    share_surveys = query_share_survey_by_batch_no(batch_no, session)
+    if len(share_surveys) == 0:
+        raise NotFound('Batch number does not exist')
+    share_surveys_list = [share_survey.to_share_dict() for share_survey in share_surveys]
+    respondent = get_respondent_by_email(share_surveys_list[0]['email_address'])
+    new_respondent = query_respondent_by_party_uuid(respondent['id'], session)
+
+    for pending_share_survey in share_surveys_list:
+        business_id = pending_share_survey['business_id']
+        survey_id = pending_share_survey['survey_id']
+        business_respondent = query_business_respondent_by_respondent_id_and_business_id(
+            business_id, new_respondent.id, session)
+        if not business_respondent:
+            # Associate respondent with new business
+            business = query_business_by_party_uuid(business_id, session)
+            if not business:
+                logger.error("Could not find business", business_id=business_id)
+                raise InternalServerError("Could not locate business when creating business association")
+            business_respondent = BusinessRespondent(business=business, respondent=new_respondent)
+        if not is_already_enrolled(survey_id, new_respondent.id, business_id, session):
+            try:
+                with session.begin_nested():
+                    enrolment = Enrolment(business_respondent=business_respondent,
+                                          survey_id=pending_share_survey['survey_id'],
+                                          status=EnrolmentStatus.ENABLED)
+                    session.add(enrolment)
+
+            except SQLAlchemyError as e:
+                logger.exception('Unable to confirm pending share survey', batch_no=batch_no)
+                session.rollback()
+                raise  # re raise the exception aimed at the generic handler
+        else:
+            logger.info('Ignoring respondent as already enrolled', business_id=business_id, survey_id=survey_id,
+                        email=share_surveys_list[0]['email_address'])
+        delete_share_survey_by_batch_no(batch_no, session)
+        session.commit()
```

Comment:
i don't think you need this commit as its in the @with_db_session wrapper

Reply:
👍